### PR TITLE
Filter Posts by Covered

### LINKED ISSUE #167: [Feed] [BE] Filter posts by covered/uncovered
## Context

We want to allow admins and content creators to be able to show 3 possible states of posts: only covered posts, only uncovered posts, and covered and uncovered posts. This is to enable users to quickly find a post to delete or uncover and to see the posts that volunteers are only able to view. 

## Requirements

You are to modify `post.getFilteredPosts` to take in an additional, **optional** input of `covered`, a boolean value. Describing our 3 states:
- only covered posts (input `covered: true`)
- only uncovered posts (input `covered: false`)
- covered and uncovered posts (covered not in input)

The result of the query should correspond correctly the states above. Write integration tests validating your implementation.

#### PM

- [ ] All 3 states are retrievable
- [ ] They are still in chronological order

#### EM

- [ ] Modified query fetches our 3 states correctly
- [ ] Integration tests written for all 3 states
- [ ] Comment describing behavior of new input `covered`

### PR DESCRIPTION
**Pull request by uma-anand**
## Filter Posts by Covered

Issue Number(s): #167.

What does this PR change and why?
We want to allow admins and content creators to be able to show 3 possible states of posts: only covered posts, only uncovered posts, and covered and uncovered posts. This is to enable users to quickly find a post to delete or uncover and to see the posts that volunteers are only able to view.

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [ ] Code follows design and style guidelines
- [ ] Commits follow guidelines (concise, squashed, etc)
- [ ] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

- Database change/migration to run
Call getFilteredPost with object instead of with a parameter. See details of implementation.

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

File: web/server/routers/post.ts [hunk 1]
```diff
@@ -325,12 +331,12 @@ export const postRouter = router({
         }
       }, {});
 
-      let completeFilter = {
-        breed: { $in: input.breed },
-        type: { $in: input.type },
-        age: { $in: input.age },
-        size: { $in: input.size },
-        gender: { $in: input.gender },
+      let noncoveredFilter = {
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Change this to `const` - rename to something generic

File: web/server/routers/post.ts [hunk 2]
```diff
@@ -350,6 +356,12 @@ export const postRouter = router({
         houseTrained: { $in: houseTrained },
         spayNeuterStatus: { $in: spayNeuterStatus },
       };
+      let completeFilter;
+      if (input.covered != undefined) {
+        completeFilter = { ...noncoveredFilter, covered: input.covered };
+      } else {
+        completeFilter = noncoveredFilter;
+      }
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Change this to:
```js
if (input.covered !== undefined) {
  completeFilter.covered = input.covered;
}
```
since reassignment isn't necessary here.

File: web/server/routers/post.ts [hunk 3]
```diff
@@ -350,6 +356,12 @@ export const postRouter = router({
         houseTrained: { $in: houseTrained },
         spayNeuterStatus: { $in: spayNeuterStatus },
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Could you go ahead and remove these fields in this PR? There was a change requested by the client and we now want to remove the filter ability for house trained and spay/neuter status.

File: web/server/routers/post.ts [hunk 4]
```diff
@@ -293,16 +293,22 @@ export const postRouter = router({
       return getAttachments(input._id);
     }),
   getFilteredPosts: procedure
-    .input(postFilterSchema)
+    .input(
+      z.object({
+        postFilterSchema: postFilterSchema,
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Update `postFilterSchema` with the above change

File: web/server/routers/post.ts [hunk 5]
```diff
@@ -293,16 +293,22 @@ export const postRouter = router({
...
-        ? [input.houseTrained]
+      const postFilterSchema = input.postFilterSchema;
+      const houseTrained = postFilterSchema.houseTrained
+        ? [postFilterSchema.houseTrained]
         : Object.values(Trained);
-      const spayNeuterStatus = input.spayNeuterStatus
-        ? [input.spayNeuterStatus]
+      const spayNeuterStatus = postFilterSchema.spayNeuterStatus
+        ? [postFilterSchema.spayNeuterStatus]
         : Object.values(Trained);
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Remove

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 6]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      jest.restoreAllMocks();
+    });
+
+    afterEach(async () => {
+      jest.restoreAllMocks();
+      await Post.deleteMany({});
+    });
+
+    test("singleton covered-uncovered", async () => {
+      const coveredPost = createRandomPost();
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Rename to just `post`

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 7]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      coveredPost["covered"] = true;
+      let id = new mongoose.Types.ObjectId();
+      const createCoveredPost = await Post.create({
+        ...coveredPost,
+        _id: id,
+        attachments,
+      });
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: You can specify the value of `covered` within this object:
```js
const coveredPost = await Post.create({
  ...post,
  _id: coveredId,
  covered: true
})
```

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 8]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+  describe("post.getFilteredPosts", () => {
+    let caller: ReturnType<typeof appRouter.createCaller>;
+    const attachmentBodyMap = {
+      "no-resize.png": noResizeData,
+      "video.mp4": videoData,
+    } as const;
+    const attachments: Array<keyof typeof attachmentBodyMap> = [
+      "no-resize.png",
+      "video.mp4",
+    ];
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: You can remove this - the `getFilteredPosts` logic shouldn't interfere with the attachments

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 9]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+        attachments,
+      });
+      expect(createUncoveredPost).not.toBeNull();
+      let goodWithArray = [];
+      for (const goodWithKey in goodWithMap) {
+        if (createCoveredPost[goodWithKey] == "yes") {
+          goodWithArray.push(goodWithMap[goodWithKey]);
+        }
+      }
+      const postFilterSchema = {
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Rename to `postFilters` or a name without schema. We should reserve "schema" for defining the structure of input since we're only passing this in as a parameter

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 10]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+
+    afterEach(async () => {
+      jest.restoreAllMocks();
+      await Post.deleteMany({});
+    });
+
+    test("singleton covered-uncovered", async () => {
+      const coveredPost = createRandomPost();
+      coveredPost["covered"] = true;
+      let id = new mongoose.Types.ObjectId();
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Keep track of which document id corresponds to the covered and uncovered, you should add an `expect` that tests whether you find those specific ids after `getFilteredPosts`

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 11]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      const createCoveredPost = await Post.create({
+        ...coveredPost,
+        _id: id,
+        attachments,
+      });
+      expect(createCoveredPost).not.toBeNull();
+      coveredPost["covered"] = false;
+      id = new mongoose.Types.ObjectId();
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: See above

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 12]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      }
+      const postFilterSchema = {
+        type: [coveredPost.type],
+        breed: coveredPost.breed,
+        age: [coveredPost.age],
+        size: [coveredPost.size],
+        gender: [coveredPost.gender],
+        behavioral: coveredPost.behavioral,
+        goodWith: goodWithArray,
+      };
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Since it's guaranteed that our collection will only have 2 documents, you should filter by the default post filters instead (like in the frontend). They look like
```js
type: array of all types
breed: array of all breeds
age: array of all ages
size: array of all sizes
gender: array of all genders
behavioral: array of all behavioral
goodWith: []
```

This guarantees that the filters will return all documents in the collection.

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 13]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+        postFilterSchema: postFilterSchema,
+      });
+      expect(filteredPosts).not.toBeNull();
+      expect(filteredPosts.length).toBe(2);
+      const coveredPosts = await caller.post.getFilteredPosts({
+        postFilterSchema: postFilterSchema,
+        covered: true,
+      });
+      expect(coveredPosts).not.toBeNull();
+      expect(coveredPosts.length).toBe(1);
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts:214]: Add a check for the coveredId

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 14]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      const uncoveredPosts = await caller.post.getFilteredPosts({
+        postFilterSchema: postFilterSchema,
+        covered: false,
+      });
+      expect(uncoveredPosts).not.toBeNull();
+      expect(uncoveredPosts.length).toBe(1);
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts:221]: Add a check for the uncoveredId

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 15]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+    });
+
+    test("random covered-uncovered", async () => {
+      let randomPost = createRandomPost();
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: You can update `createRandomPost` to randomly assign posts as covered or uncovered and remove the logic below

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 16]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+          numCovered++;
+        }
+        const id = new mongoose.Types.ObjectId();
+        createPost = await Post.create({
+          ...randomPost,
+          _id: id,
+          attachments,
+        });
+        expect(createPost).not.toBeNull();
+      }
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Just generate 10 posts, don't worry about the distribution of covered and uncovered. It's unlikely that faker will give you 10 of one and 0 of the other or something like that.

Do this in `__mocks__/post.ts`
```js
export const randomPosts = Array.from({ length: 10 }).map(() =>
  createRandomPost()
);
```

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 17]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      const postFilterSchema = {
+        type: [randomPost.type],
+        breed: randomPost.breed,
+        age: [randomPost.age],
+        size: [randomPost.size],
+        gender: [randomPost.gender],
+        behavioral: randomPost.behavioral,
+        goodWith: goodWithArray,
+      };
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: See above, use defaults

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 18]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+      const allPosts = await caller.post.getFilteredPosts({
+        postFilterSchema: postFilterSchema,
+      });
+      expect(allPosts).not.toBeNull();
+      expect(allPosts.length).toBe(numPosts);
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Instead of `numPosts` now, go off of `randomPosts`

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 19]
```diff
@@ -145,4 +155,138 @@ describe("[API] Post - Integration Test", () => {
...
+        covered: true,
+      });
+      expect(coveredPosts).not.toBeNull();
+      expect(coveredPosts.length).toBe(numCovered);
+      const uncoveredPosts = await caller.post.getFilteredPosts({
+        postFilterSchema: postFilterSchema,
+        covered: false,
+      });
+      expect(uncoveredPosts).not.toBeNull();
+      expect(uncoveredPosts.length).toBe(numUncovered);
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Use an array method to get the arrays of both covered and uncovered and use `expect(____).toMatchObject()`. This will match the two arrays and their nested values. We need this to correctly validate our implementation since the lengths alone are not enough (you should still keep the length checks and the null checks).

For the array method, you can look at how it's done in `__tests__/db/actions/User.unit.test.ts:62`.

**Review by katsukixyz — CHANGES_REQUESTED** [review]

File: web/server/routers/post.ts (hunk 2 above)
Comment (uma-anand) [reply web/server/routers/post.ts]: Sorry, I don't quite understand. Shouldn't the filter have the remaining values of the base filter? I did not directly add the covered to the base filter because that gives me a TypeError.

File: web/server/routers/post.ts [hunk 20]
```diff
@@ -353,14 +344,12 @@ export const postRouter = router({
         getsAlongWithYoungKids: {
           $in: getsAlongWith["getsAlongWithYoungKids"],
         },
-        houseTrained: { $in: houseTrained },
-        spayNeuterStatus: { $in: spayNeuterStatus },
       };
       let completeFilter;
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Sorry, I should have specified. You don't need to declare another variable here. Just work with `baseFilter`.

File: web/server/routers/post.ts [hunk 21]
```diff
@@ -353,14 +344,12 @@ export const postRouter = router({
...
-      if (input.covered != undefined) {
-        completeFilter = { ...noncoveredFilter, covered: input.covered };
+      if (input.covered !== undefined) {
+        completeFilter = { ...baseFilter, covered: input.covered };
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: `baseFilter.covered = input.covered`

File: web/server/routers/post.ts [hunk 22]
```diff
@@ -353,14 +344,12 @@ export const postRouter = router({
...
       } else {
-        completeFilter = noncoveredFilter;
+        completeFilter = baseFilter;
       }
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: You can rid of this case

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 23]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
 
     test("random covered-uncovered", async () => {
-      let randomPost = createRandomPost();
-      const numPosts = Math.ceil(Math.random() * 10);
-      let numCovered = 0;
-      let numUncovered = 0;
+      const randPosts = randomPosts;
+      let allIds = [];
+      let covered = [];
+      let uncovered = [];
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: You don't need to maintain an array for ids, covered, and uncovered posts - just have `expect().toMatchObject()` with the filtered arrays handle the checking

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 24]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
       expect(uncoveredPosts.length).toBe(1);
+      expect(uncoveredPosts[0]._id).toMatchObject(uncoveredId);
     });
 
     test("random covered-uncovered", async () => {
-      let randomPost = createRandomPost();
-      const numPosts = Math.ceil(Math.random() * 10);
-      let numCovered = 0;
-      let numUncovered = 0;
+      const randPosts = randomPosts;
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Remove this, you can directly reference `randomPosts` in your expects

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 25]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
+          ? covered.push(id.toString())
+          : uncovered.push(id.toString());
         createPost = await Post.create({
-          ...randomPost,
+          ...randPosts[i],
           _id: id,
-          attachments,
         });
         expect(createPost).not.toBeNull();
       }
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Use `Post.insertMany` instead of looping 10 times and individually adding the documents

File: web/server/routers/post.ts [hunk 26]
```diff
@@ -301,12 +298,6 @@ export const postRouter = router({
     )
```
Comment (katsukixyz) [inline web/server/routers/post.ts:299]: Rename the `postFilterSchema` key to just `postFilters`

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 27]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
       };
       const allPosts = await caller.post.getFilteredPosts({
-        postFilterSchema: postFilterSchema,
+        postFilterSchema: postFilters,
       });
       expect(allPosts).not.toBeNull();
-      expect(allPosts.length).toBe(numPosts);
+      expect(allPosts.length).toBe(randPosts.length);
+      let expected = allPosts.map(({ _id, ...user }) => _id.toString());
+      expect(expected).toEqual(expect.arrayContaining(allIds));
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: Can you directly match `allPosts` to `randomPosts` instead (`toMatchObject`)?

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 28]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
       const coveredPosts = await caller.post.getFilteredPosts({
-        postFilterSchema: postFilterSchema,
+        postFilterSchema: postFilters,
         covered: true,
       });
       expect(coveredPosts).not.toBeNull();
-      expect(coveredPosts.length).toBe(numCovered);
+      expect(coveredPosts.length).toBe(covered.length);
+      expected = coveredPosts.map(({ _id, ...user }) => _id.toString());
+      expect(expected).toEqual(expect.arrayContaining(covered));
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: To be sure of the implementation, we need to check the entire array all fields within the objects.

```js
const expectedCoveredPosts = randomPosts.filter((post) => post.covered);
expect(coveredPosts).toMatchObject(expectedCoveredPosts);
```

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 29]
```diff
@@ -181,112 +183,101 @@ describe("[API] Post - Integration Test", () => {
...
       const uncoveredPosts = await caller.post.getFilteredPosts({
-        postFilterSchema: postFilterSchema,
+        postFilterSchema: postFilters,
         covered: false,
       });
       expect(uncoveredPosts).not.toBeNull();
-      expect(uncoveredPosts.length).toBe(numUncovered);
+      expect(uncoveredPosts.length).toBe(uncovered.length);
+      expected = uncoveredPosts.map(({ _id, ...user }) => _id.toString());
+      expect(expected).toEqual(expect.arrayContaining(uncovered));
```
Comment (katsukixyz) [inline web/__tests__/pages/api/post/post.int.test.ts]: See above

**Review by katsukixyz — CHANGES_REQUESTED** [review]

**Comment by uma-anand** [comment]
I couldn't get the toMatchObject to work as mentioned so I've used another way of doing it but it now does make sure all the fields are the same and not just the ID.

**Review by katsukixyz — APPROVED** [review]
